aiidalab_widgets_base/wizard.py

```diff
@@ -0,0 +1,212 @@
+"""The wizard application allows the implication of a Wizard-like GUI.
+
+Authors:
+
+    * Carl Simon Adorf <simon.adorf@epfl.ch>
+"""
+from enum import Enum
+from time import sleep
+from time import time
+from threading import Thread
+
+import traitlets
+import ipywidgets as ipw
+
+
+class WizardAppWidgetStep(traitlets.HasTraits):
+    "One step of a WizardAppWidget."
+
+    class State(Enum):
+        INIT = 0  # implicit default value
+
+        CONFIGURED = 1
+        READY = 2
+        ACTIVE = 3
+        SUCCESS = 4
+
+        # All error states have negative codes
+        FAIL = -1
+
+    state = traitlets.UseEnum(State)
+    auto_next = traitlets.Bool()
+
+    def can_reset(self):
+        return hasattr(self, "reset")
+
+
+class WizardAppWidget(ipw.VBox):
+    ICON_SEPARATOR = "\u2000"  # en-dash  (placed between title and icon)
+
+    ICONS = {
+        WizardAppWidgetStep.State.INIT: "\u25cb",
+        WizardAppWidgetStep.State.READY: "\u25ce",
+        WizardAppWidgetStep.State.CONFIGURED: "\u25cf",
+        WizardAppWidgetStep.State.ACTIVE: ["\u25dc", "\u25dd", "\u25de", "\u25df"],
+        WizardAppWidgetStep.State.SUCCESS: "\u2713",
+        WizardAppWidgetStep.State.FAIL: "\u00d7",
+    }
+
+    @classmethod
+    def icons(cls):
+        """Return the icon set and return animated icons based on the current time stamp."""
+        t = time()
+        return {
+            key: item
+            if isinstance(item, str)
+            else item[int((t * len(item) % len(item)))]
+            for key, item in cls.ICONS.items()
+        }
+
+    selected_index = traitlets.Int(allow_none=True)
+
+    def __init__(self, steps, **kwargs):
+        # The number of steps must be greater than one
+        # for this app's logic to make sense.
+        assert len(steps) > 1
+        self.steps = steps
+
+        # Unzip the steps to titles and widgets.
+        self.titles, widgets = zip(*steps)
+
+        # Initialize the accordion with the widgets ...
+        self.accordion = ipw.Accordion(children=widgets)
+        self._update_titles()
+        ipw.link((self.accordion, "selected_index"), (self, "selected_index"))
+
+        # Automatically update titles to implement the "spinner"
+
+        def spinner_thread():
+            while True:
+                sleep(0.1)
+                self._update_titles()
+
+        Thread(target=spinner_thread).start()
+
+        # Watch for changes to each step's state
+        for widget in widgets:
+            assert widget.has_trait("state")
```

Comment:
Explicit type checks are not in the spirit of Python's duck-typing (dynamic typing) paradigm and unless really needed should be avoided as it poses an unnecessary constraint.

Reply:
Pragmatism vs. idealism :)
In any case, `assert` shouldn't be used here, as it's considered a statement for debugging and testing, e.g. all `assert` statements can be subverted running the python interpreter with the `-O` option.

Reply:
Ok, then let's look at this from a slightly different angle. To make sure things work properly, do we want to require that widgets are derived from `WizardAppWidgetStep`? Or should we allow widgets that are not derived from it?

Reply:
We are explicitly observing that `state` trait in the next line it would fail either way (even with `-O` option...), the assertion could just as well be removed entirely. I don't really see a point why one would not use a `WizardAppWidgetStep` to implement a step, but I also see no real point in making that a hard requirement. To make sure that apps work, developers must test them. If they forget to implement the `step` interface in their steps, the app is going to fail. You could use the same argument for or against type checks in the rest of the language.

Reply:
> To make sure that apps work, developers must test them. If they forget to implement the `step` interface in their steps, the app is going to fail.

My advice on the explicit error was actually to help developers in writing their widgets. Additionally, requiring the widgets to derive from `WizardAppWidgetStep` would help in ensuring backwards compatibility. Imagine the relations between `WizardAppWidget` - `WizardAppWidgetStep` change. If a developer wasn't using the `WizardAppWidgetStep` as a base class - things might not be working anymore for them.

Reply:
I understand that, but the documentation is helping developers already with that. We are showing that you should derive your steps from this class and things are going to immediately fail if a developers forgets to do so. I could provide a more detailed error message here, something like:

> The step widget has no `.state` trait, did you forget to derive the step from the `WizardAppWidgetStep` class?

However, I just don't see a good reason to stray from core language paradigms here.

> Additionally, requiring the widgets to derive from WizardAppWidgetStep would help in ensuring backwards compatibility. That does not really have anything to do with this problem IMO.

Again, developers should absolute derive from that class in 99% of cases – this is not in question. I just don't really a good reason to *force* them to.